### PR TITLE
LPS-189291 New type of Feature Flag: deprecation

### PR DESCRIPTION
**Pull request by thiago-buarqque**
Related to story: [LPS-189291](https://liferay.atlassian.net/browse/LPS-189291)

### Goal
Add a new type of Feature Flag called `deprecation` that must have its value set as `true` (appears for the user) by default. For a new company the default value should be `false`.

### Additional info
- Adding a integration test to assert that a new company has its deprecation feature flags value as false, I had to add the feature-flag-api and -impl modules, setup -test module and refactor -web module.

### CODE REVIEW

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
Hey @drewbrokke @ethib137 , could you guys please review this PR?

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties
Comment (ethib137) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:7020]: The translation should be "Deprecation".

**Review by ethib137 — COMMENTED** [review]
@thiago-buarqque , this is looking good. Don't worry about replying to all my comments. Let's have a call today to discuss the PR. I'm sure with some of my comments I'm missing some context that likely invalidates my recommendation. Thanks for the work on this!

File: modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java
Comment (thiago-buarqque) [reply modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java:113]: I need the index variables (`i`) to insert into the `disabledFeatureFlagKeys` array, that's why it's not a for each.

File: modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java
Comment (thiago-buarqque) [reply modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java:116]: You're right. I'm going to change that

File: modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java
Comment (thiago-buarqque) [reply modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java:119]: @drewbrokke, the disabledFeatureFlagKeys property name is good considering we're getting it from [DeprecationFeatureFlags.java](https://github.com/liferay-peds/liferay-portal/pull/205/commits/e03dc3b9b89b9108319115a4710b51a7bcb5ec82#diff-d4156ce1e7ff5e2697867e275308792db37f2bdf242a8d810888319463e5e8a0) configuration. In a pseudo level, something like:

```
configuration = deprecationFeatureFlags.getConfiguration();
configuration.getProperty("deprecationFeatureFlagKeys")
```
Does it make sense to change it to match the new array name `deprecationFeatureFlagKeys`?

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties
Comment (thiago-buarqque) [reply modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:7020]: When talking to Mateo, he said it should be `Deprecated`. That was last month when I started this story. Either that changed or he just missed that. I'll change it. Thank you.

Comment (thiago-buarqque) [reply modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:7020]: My bad

File: modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java
Comment (thiago-buarqque) [reply modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java:88]: It does. I'm kind of unaware regarding the LXC. But wouldn't each customer have a DXP code copy (or something similar)? O are we having a single Liferay DXP running for all clients? 

The solution here should work fine considering a instance for each customer. The `activate` ([see declaration](https://github.com/liferay-peds/liferay-portal/pull/205/files/2e18f0d0ca7b8eddf3791cbe5725ea0eb28bdb04#diff-0e5dc64f4c46e9237e6b30e7c2bd39f45ba1e6dae9c1efda9162d5281f72b449R78)) method runs only when the portal is starting (always before the `onAfterCreate` method). Then when creating the companies, we read the configuration created in there.

Am I missing something?
cc: @drewbrokke

**Comment by eltonccdantas** [comment]
ci:reevaluate:1257401_2341

**Comment by thiago-buarqque** [comment]
@eltonccdantas the errors are on me. I was king of expecting that, currently fixing those. Thanks

**Comment by eltonccdantas** [comment]
> @eltonccdantas the errors are on me. I was king of expecting that, currently fixing those. Thanks

Alrighty, feel free to tag me when needed.

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties
Comment (ethib137) [reply modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:7020]: Nope, the description should still be "Deprecated features". Maybe this will help... When we deprecate a feature, we use a Deprecation feature flag to toggle the visibility of the deprecated feature.

File: modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java
Comment (ethib137) [reply modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java:119]: After switching to use the current deprecated features whenever adding a new company, will we even still need to add this property to the properties map?

Comment (thiago-buarqque) [reply modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagsCompanyModelListener.java:119]: No, I forgot to mention that here, sorry

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Review by ethib137 — APPROVED** [review]
Looks good. I like how much simpler the code is now. Just a couple comments to consider.

File: modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagCompanyModelListener.java
Comment (thiago-buarqque) [reply modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagCompanyModelListener.java:41]: You're right, changing it to featureFlag

File: modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagCompanyModelListener.java
Comment (drewbrokke) [reply modules/apps/feature-flag/feature-flag-impl/src/main/java/com/liferay/feature/flag/internal/model/listener/DeprecationFeatureFlagCompanyModelListener.java:33]: I might consider renaming the entire `CompanyFeatureFlags` class at this point (especially while it's still internal). Perhaps even if we just called it `FeatureFlags`, we would have flexibility in variable naming going forward.

File: modules/apps/feature-flag/feature-flag-api/src/main/java/com/liferay/feature/flag/manager/FeatureFlagPreferencesManager.java
Comment (drewbrokke) [inline modules/apps/feature-flag/feature-flag-api/src/main/java/com/liferay/feature/flag/manager/FeatureFlagPreferencesManager.java:11]: Does this need to be part of the API? I think this could stay internal to the `impl` module, right?

File: modules/apps/feature-flag/feature-flag-api/src/main/java/com/liferay/feature/flag/model/FeatureFlagDisplay.java
Comment (drewbrokke) [inline modules/apps/feature-flag/feature-flag-api/src/main/java/com/liferay/feature/flag/model/FeatureFlagDisplay.java:6]: Why does this need to be API?

**Review by drewbrokke — CHANGES_REQUESTED** [review]
Hey all, I did not comment on them all, but I'm seeing quite a few classes being moved into the API module that IMO should not be there. We should aim to keep the API as small as possible. For example, all the implementations of `FeatureFlag` really don't need to be API, only the `FeatureFlag` interface. And the preferences manager is an implementation detail used by the CompanyFeatureFlags.

The strategy should be to keep as much in the `web` and `impl` modules as possible. It's very easy to add new API, but not so easy to change or remove stuff once it's there.

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:stop:relevant

**Comment by thiago-buarqque** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by thiago-buarqque** [comment]
ci:test:rsf

**Comment by thiago-buarqque** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by john-co** [comment]
Reviewed. Test failures are unrelated :heavy_check_mark: 

Ok to forward from QA perspective.

**Comment by john-co** [comment]
ci:forward:force

**Comment by thiago-buarqque** [comment]
Hey @john-co, this PR is still waiting for a review from @drewbrokke, I believe he was waiting CI to be fixed so he could review it

**Comment by thiago-buarqque** [comment]
ci:forward:stop

**Comment by eltonccdantas** [comment]
Hey @thiago-buarqque, I closed it to impede it to go forward. When CI stop trying forwarding it, we reopen it, though.

**Comment by thiago-buarqque** [comment]
Thanks @eltonccdantas, I resent the PR to #216 , this one is messy, lots of CI comments.

**Pull request closed without merging**
